### ISSUE TITLE
`esp-idf-sys` fails to build

### ISSUE THREAD

**Issue by 999eagle** [esp-rs/esp-idf-sys#49]
I have a project using the Rust ESP toolchain and today compilation of `esp-idf-sys` failed without any apparent changes. I've created an entirely new project from scratch using https://github.com/esp-rs/esp-idf-template and even that fails to compile with the exact same error:

```
    Updating crates.io index
   Compiling compiler_builtins v0.1.49
   Compiling core v0.0.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/core)
   Compiling libc v0.2.103
   Compiling libc v0.2.112
   Compiling cc v1.0.69
   Compiling proc-macro2 v1.0.36
   Compiling std v0.0.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/std)
   Compiling unicode-xid v0.2.2
   Compiling cfg-if v1.0.0
   Compiling syn v1.0.84
   Compiling memchr v2.4.1
   Compiling cc v1.0.72
   Compiling serde_derive v1.0.133
   Compiling fnv v1.0.7
   Compiling log v0.4.14
   Compiling once_cell v1.9.0
   Compiling autocfg v1.0.1
   Compiling regex-syntax v0.6.25
   Compiling untrusted v0.7.1
   Compiling spin v0.5.2
   Compiling serde v1.0.133
   Compiling adler v1.0.2
   Compiling version_check v0.9.4
   Compiling glob v0.3.0
   Compiling tinyvec_macros v0.1.0
   Compiling matches v0.1.9
   Compiling lazy_static v1.4.0
   Compiling crossbeam-utils v0.8.5
   Compiling crc32fast v1.3.0
   Compiling gimli v0.26.1
   Compiling minimal-lexical v0.2.1
   Compiling bitflags v1.3.2
   Compiling unicode-bidi v0.3.7
   Compiling same-file v1.0.6
   Compiling unicode-width v0.1.9
   Compiling percent-encoding v2.1.0
   Compiling ppv-lite86 v0.2.16
   Compiling strsim v0.8.0
   Compiling rustc-demangle v0.1.21
   Compiling serde_json v1.0.74
   Compiling humantime v2.1.0
   Compiling ansi_term v0.12.1
   Compiling termcolor v1.1.2
   Compiling vec_map v0.8.2
   Compiling anyhow v1.0.52
   Compiling bindgen v0.59.2
   Compiling either v1.6.1
   Compiling remove_dir_all v0.5.3
   Compiling ryu v1.0.9
   Compiling rustc-hash v1.1.0
   Compiling strsim v0.10.0
   Compiling peeking_take_while v0.1.2
   Compiling unicode-segmentation v1.8.0
   Compiling zero v0.1.2
   Compiling ident_case v1.0.1
   Compiling chunked_transfer v1.4.0
   Compiling shlex v1.1.0
   Compiling lazycell v1.3.0
   Compiling base64 v0.13.0
   Compiling itoa v1.0.1
   Compiling async-trait v0.1.52
   Compiling paste v1.0.6
   Compiling libloading v0.7.2
   Compiling tinyvec v1.5.1
   Compiling thread_local v1.1.3
   Compiling walkdir v2.3.2
   Compiling textwrap v0.11.0
   Compiling form_urlencoded v1.0.1
   Compiling nom v7.1.0
   Compiling xmas-elf v0.8.0
   Compiling miniz_oxide v0.4.4
   Compiling clang-sys v1.3.0
   Compiling heck v0.3.3
   Compiling quote v1.0.14
   Compiling aho-corasick v0.7.18
   Compiling bstr v0.2.17
   Compiling object v0.27.1
   Compiling ring v0.16.20
   Compiling backtrace v0.3.63
   Compiling unwind v0.0.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/unwind)
   Compiling getrandom v0.2.3
   Compiling atty v0.2.14
   Compiling which v4.2.2
   Compiling dirs-sys v0.3.6
   Compiling flate2 v1.0.22
   Compiling unicode-normalization v0.1.19
   Compiling clap v2.34.0
   Compiling rand_core v0.6.3
   Compiling dirs v4.0.0
   Compiling rand_chacha v0.3.1
   Compiling rand v0.8.4
   Compiling regex v1.5.4
   Compiling idna v0.2.3
   Compiling tempfile v3.2.0
   Compiling addr2line v0.17.0
   Compiling url v2.2.2
   Compiling cexpr v0.6.0
   Compiling globset v0.4.8
   Compiling env_logger v0.9.0
   Compiling ignore v0.4.18
   Compiling webpki v0.22.0
   Compiling sct v0.7.0
   Compiling webpki-roots v0.22.2
   Compiling rustls v0.20.2
   Compiling globwalk v0.8.1
   Compiling darling_core v0.13.1
   Compiling ureq v2.4.0
   Compiling strum_macros v0.21.1
   Compiling darling_macro v0.13.1
   Compiling darling v0.13.1
   Compiling enumset_derive v0.5.5
   Compiling strum v0.21.0
   Compiling rustc-std-workspace-core v1.99.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/rustc-std-workspace-core)
   Compiling alloc v0.0.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/alloc)
   Compiling cfg-if v0.1.10
   Compiling rustc-std-workspace-alloc v1.99.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/rustc-std-workspace-alloc)
   Compiling panic_abort v0.0.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/panic_abort)
   Compiling panic_unwind v0.0.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/panic_unwind)
   Compiling hashbrown v0.11.0
   Compiling std_detect v0.1.5 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/stdarch/crates/std_detect)
   Compiling toml v0.5.8
   Compiling proc_macro v0.0.0 (/home/sophie/.rustup/toolchains/esp/lib/rustlib/src/rust/library/proc_macro)
   Compiling cargo_toml v0.9.2
   Compiling base64 v0.12.3
   Compiling no-std-net v0.5.0
   Compiling enumset v1.0.8
   Compiling mutex-trait v0.2.0
   Compiling http-auth-basic v0.1.3
   Compiling embuild v0.27.1
   Compiling esp-idf-sys v0.28.3
   Compiling esp-test v0.1.0 (/home/sophie/projects/other/esp-test)
   Compiling embedded-svc v0.15.4
error: failed to run custom build command for `esp-idf-sys v0.28.3`

Caused by:
  process didn't exit successfully: `/home/sophie/projects/other/esp-test/target/debug/build/esp-idf-sys-e4ce039c35a62e6f/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=ESP_IDF_TOOLS_INSTALL_DIR
  cargo:rerun-if-env-changed=ESP_IDF_SDKCONFIG
  cargo:rerun-if-env-changed=ESP_IDF_SDKCONFIG_DEFAULTS
  cargo:rerun-if-env-changed=MCU
  Installer version: 1.0.2
  Platform: Linux-5.15.11-zen1-1-zen-x86_64-with-glibc2.33
  Python version: 3.10.1 (main, Dec 18 2021, 23:53:45) [GCC 11.1.0]
  Python path: /usr/bin/python3
  Creating a virtual environment at /home/sophie/projects/other/esp-test/.embuild/platformio/penv
  Updating Python package manager (PIP) in a virtual environment
  PIP has been successfully updated!
  Virtual environment has been successfully created!
  Installing PlatformIO Core
  Collecting platformio
    Using cached platformio-5.2.4-py3-none-any.whl
  Collecting requests==2.*
    Using cached requests-2.26.0-py2.py3-none-any.whl (62 kB)
  Collecting semantic-version==2.8.*
    Using cached semantic_version-2.8.5-py2.py3-none-any.whl (15 kB)
  Collecting colorama
    Using cached colorama-0.4.4-py2.py3-none-any.whl (16 kB)
  Collecting starlette==0.17.*
    Using cached starlette-0.17.1-py3-none-any.whl (58 kB)
  Collecting pyelftools<1,>=0.27
    Using cached pyelftools-0.27-py2.py3-none-any.whl (151 kB)
  Collecting aiofiles==0.8.*
    Using cached aiofiles-0.8.0-py3-none-any.whl (13 kB)
  Collecting ajsonrpc==1.*
    Using cached ajsonrpc-1.2.0-py3-none-any.whl (22 kB)
  Collecting tabulate==0.8.*
    Using cached tabulate-0.8.9-py3-none-any.whl (25 kB)
  Collecting uvicorn==0.16.*
    Using cached uvicorn-0.16.0-py3-none-any.whl (54 kB)
  Collecting bottle==0.12.*
    Using cached bottle-0.12.19-py3-none-any.whl (89 kB)
  Collecting marshmallow<4,>=2
    Using cached marshmallow-3.14.1-py3-none-any.whl (47 kB)
  Collecting zeroconf==0.37.*
    Using cached zeroconf-0.37.0-py3-none-any.whl (105 kB)
  Collecting click!=8.0.2,<9,>=8
    Using cached click-8.0.3-py3-none-any.whl (97 kB)
  Collecting pyserial==3.*
    Using cached pyserial-3.5-py2.py3-none-any.whl (90 kB)
  Collecting wsproto==1.0.*
    Using cached wsproto-1.0.0-py3-none-any.whl (24 kB)
  Collecting idna<4,>=2.5
    Using cached idna-3.3-py3-none-any.whl (61 kB)
  Collecting charset-normalizer~=2.0.0
    Using cached charset_normalizer-2.0.9-py3-none-any.whl (39 kB)
  Collecting urllib3<1.27,>=1.21.1
    Using cached urllib3-1.26.7-py2.py3-none-any.whl (138 kB)
  Collecting certifi>=2017.4.17
    Using cached certifi-2021.10.8-py2.py3-none-any.whl (149 kB)
  Collecting anyio<4,>=3.0.0
    Using cached anyio-3.4.0-py3-none-any.whl (78 kB)
  Collecting h11>=0.8
    Using cached h11-0.12.0-py3-none-any.whl (54 kB)
  Collecting asgiref>=3.4.0
    Using cached asgiref-3.4.1-py3-none-any.whl (25 kB)
  Collecting ifaddr>=0.1.7
    Using cached ifaddr-0.1.7-py2.py3-none-any.whl (10 kB)
  Collecting sniffio>=1.1
    Using cached sniffio-1.2.0-py3-none-any.whl (10 kB)
  Installing collected packages: sniffio, idna, urllib3, ifaddr, h11, click, charset-normalizer, certifi, asgiref, anyio, zeroconf, wsproto, uvicorn, tabulate, starlette, semantic-version, requests, pyserial, pyelftools, marshmallow, colorama, bottle, ajsonrpc, aiofiles, platformio
  Successfully installed aiofiles-0.8.0 ajsonrpc-1.2.0 anyio-3.4.0 asgiref-3.4.1 bottle-0.12.19 certifi-2021.10.8 charset-normalizer-2.0.9 click-8.0.3 colorama-0.4.4 h11-0.12.0 idna-3.3 ifaddr-0.1.7 marshmallow-3.14.1 platformio-5.2.4 pyelftools-0.27 pyserial-3.5 requests-2.26.0 semantic-version-2.8.5 sniffio-1.2.0 starlette-0.17.1 tabulate-0.8.9 urllib3-1.26.7 uvicorn-0.16.0 wsproto-1.0.0 zeroconf-0.37.0

  PlatformIO Core has been successfully installed into an isolated environment `/home/sophie/projects/other/esp-test/.embuild/platformio/penv`!

  The full path to `platformio.exe` is `/home/sophie/projects/other/esp-test/.embuild/platformio/penv/bin/platformio`

  If you need an access to `platformio.exe` from other applications, please install Shell Commands
  (add PlatformIO Core binary directory `/home/sophie/projects/other/esp-test/.embuild/platformio/penv/bin` to the system environment PATH variable):

  See https://docs.platformio.org/page/installation.html#install-shell-commands

  Found compatible PlatformIO Core 5.2.4 -> /home/sophie/projects/other/esp-test/.embuild/platformio/penv/bin/platformio

  --- stderr
  Error: Compatible PlatformIO Core not found.
  Reason: PlatformIO executable not found in `/home/sophie/projects/other/esp-test/.embuild/platformio/penv/bin`
  Error: invalid type: null, expected a string at line 1 column 9390
```

Every subsequent build (without cleaning the `.embuild` directory) fails with:

```
   Compiling esp-idf-sys v0.28.3
error: failed to run custom build command for `esp-idf-sys v0.28.3`

Caused by:
  process didn't exit successfully: `/home/sophie/projects/other/esp-test/target/debug/build/esp-idf-sys-e4ce039c35a62e6f/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=ESP_IDF_TOOLS_INSTALL_DIR
  cargo:rerun-if-env-changed=ESP_IDF_SDKCONFIG
  cargo:rerun-if-env-changed=ESP_IDF_SDKCONFIG_DEFAULTS
  cargo:rerun-if-env-changed=MCU
  Found compatible PlatformIO Core 5.2.4 -> /home/sophie/projects/other/esp-test/.embuild/platformio/penv/bin/platformio

  --- stderr
  Error: invalid type: null, expected a string at line 1 column 9390
```

**Comment by 999eagle** [comment]
Update: I've tried to use `cargo-pio`. Running `cargo pio -v upgrade` in the bare bones template project results in the same error:

```
Using built-in get-platformio.py
Running command "python3" "/tmp/.tmp4gkLlO" "check" "core" "--dump-state" "/tmp/.tmpgxvBZ2"
Found compatible PlatformIO Core 5.2.4 -> /home/sophie/.platformio/penv/bin/platformio
Resolving Resolver { pio: Pio { is_develop_core: false, platformio_exe: "/home/sophie/.platformio/penv/bin/platformio", penv_dir: "/home/sophie/.platformio/penv", installer_version: "1.0.2", python_version: "3.10.1", core_version: "5.2.4", system: "linux_x86_64", python_exe: "/home/sophie/.platformio/penv/bin/python", cache_dir: "/home/sophie/.platformio/.cache", penv_bin_dir: "/home/sophie/.platformio/penv/bin", core_dir: "/home/sophie/.platformio", log_level: Verbose }, params: ResolutionParams { board: None, mcu: None, platform: None, frameworks: [], target: None } }
Running PlatformIO command "/home/sophie/.platformio/penv/bin/platformio" "platform" "frameworks" "--json-output"
Error: invalid type: null, expected a string at line 1 column 9390
```

Manually running `platformio platform frameworks --json-output` results in a huge JSON file, here's the relevant snippet:

`{"name": "tirtos", "title": "TI-RTOS", "description": "Stripped Energia framework without compiled core", "url": null, "homepage": "https://platformio.org/frameworks/tirtos", "platforms": ["timsp432"]}`

Apparently `cargo-pio` chokes on the `url` field being `null` for an entirely unrelated platform.

**Comment by mihai-dinculescu** [comment]
@999eagle your fix works for me, thanks!

Until it is merged into embuild and released, I'm using it via
```
[patch.crates-io]
embuild = { git="https://github.com/999eagle/embuild" }
```

**Comment by N3xed** [comment]
Thanks for opening this issue.
I'll keep this open till a new `embuild` version is released.

**Comment by ivmarkov** [comment]
We probably want to release a new embuild asap, as all CI builds currently fail because of this issue?

**Comment by N3xed** [comment]
Fixed with new `embuild` version `0.27.2`.
